File: backend/models/userModel.js
const mongoose = require("mongoose");
const validator = require("validator");
const bcryptjs = require("bcryptjs");
const jwt = require("jsonwebtoken");
const crypto = require("crypto");

const userSchema = new mongoose.Schema({
  name: {
    type: String,
    required: [true, "Enter your Name"],
  },
  email: {
    type: String,
    required: [true, "Enter your E-mail"],
    unique: true,
    validate: [validator.isEmail, "Please Enter a valid Email"],
  },
  password: {
    type: String,
    required: [true, "Please Enter your Password"],
    minLength: [8, "Password must be of 8 characters in minimum."],
    select: false,
  },
  profile_image: {
    public_id: {
      type: String,
      required: true,
    },
    url: {
      type: String,
      required: true,
    },
  },
  role: {
    type: String,
    default: "user",
  },
  resetPasswordToken: String,
  resetPasswordExpire: Date,
});

//to make sure password is not encrypted multiple time during other updation
userSchema.pre("save", async function (next) {
  if (this.isModified("password")) {
    this.password = await bcryptjs.hash(this.password, 10);
  }
  next();
});

//JWT Token
userSchema.methods.getJWTToken = function () {
  return jwt.sign({ id: this._id }, process.env.JWT_SECRET, {
    expiresIn: process.env.JWT_EXPIRATION,
  });
};

userSchema.methods.comparePassword = async function (enteredPassword) {
  return await bcryptjs.compare(enteredPassword, this.password);
};

//for resetting password
userSchema.methods.getResetPasswordToken = function () {
  const resetToken = crypto.randomBytes(20).toString("hex");
  //hasing and adding the token to userSchema
  this.resetPasswordToken = crypto
    .createHash("sha256")
    .update(resetToken)
    .digest("hex");
  this.resetPasswordExpire = Date.now() + 10 * 60 * 1000;
  return resetToken;
};

module.exports = mongoose.model("User", userSchema);

File: frontend/lib/data/models/userModel.dart
class User {
  String name;
  String email;
  String password;
  ProfileImage profileImage;
  String role;
  String resetPasswordToken;
  DateTime? resetPasswordExpire;

  User({
    required this.name,
    required this.email,
    required this.password,
    required this.profileImage,
    this.role = "user",
    this.resetPasswordToken = "",
    required this.resetPasswordExpire,
  });

  factory User.fromJson(Map<dynamic, dynamic> json) {
    return User(
      name: json['name'],
      email: json['email'],
      password: json['password'],
      profileImage: ProfileImage.fromJson(json['profile_image']),
      role: json['role'] ?? "user",
      resetPasswordToken: json['resetPasswordToken'] ?? "",
      resetPasswordExpire: json['resetPasswordExpire'] != null
          ? DateTime.parse(json['resetPasswordExpire'])
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'name': name,
      'email': email,
      'password': password,
      'profile_image': profileImage.toJson(),
      'role': role,
      'resetPasswordToken': resetPasswordToken,
      'resetPasswordExpire': resetPasswordExpire != null
          ? resetPasswordExpire!.toIso8601String()
          : null,
    };
  }
}

class ProfileImage {
  String publicId;
  String url;

  ProfileImage({
    required this.publicId,
    required this.url,
  });

  factory ProfileImage.fromJson(Map<String, dynamic> json) {
    return ProfileImage(
      publicId: json['public_id'],
      url: json['url'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'public_id': publicId,
      'url': url,
    };
  }
}
